
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Extractors &#8212; NexusLIMS 2.2.0.dev0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/autodoc_pydantic.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=8633d6cc" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=dc4a309c"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=30646c52"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'user_guide/extractors';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = '_static/switcher.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            true;
        </script>
    <script src="../_static/custom.js?v=ba087cdf"></script>
    <link rel="canonical" href="https://datasophos.github.io/NexusLIMS/latest/user_guide/extractors.html" />
    <link rel="icon" href="../_static/nexusLIMS_bare_logo.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="NexusLIMS Taxonomy" href="taxonomy.html" />
    <link rel="prev" title="Record building workflow" href="record_building.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="2.2.0.dev0" />
    <meta name="docbuild:last-update" content="Dec, 30, 2025"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo_horizontal_light.png" class="logo__image only-light" alt="NexusLIMS 2.2.0.dev0 documentation - Home"/>
    <img src="../_static/logo_horizontal_dark.png" class="logo__image only-dark pst-js-only" alt="NexusLIMS 2.2.0.dev0 documentation - Home"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Home
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../migration.html">
    Migration Guide: v1.4.3 to v2.2.0
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../dev_guide.html">
    Dev Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../reference.html">
    Reference
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
      
        <div class="navbar-item">

<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/datasophos/NexusLIMS" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://datasophos.co" title="Datasophos" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-solid fa-globe fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Datasophos</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Home
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../migration.html">
    Migration Guide: v1.4.3 to v2.2.0
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../dev_guide.html">
    Dev Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../reference.html">
    Reference
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/datasophos/NexusLIMS" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://datasophos.co" title="Datasophos" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-solid fa-globe fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Datasophos</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="record_building.html">Record building workflow</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Extractors</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxonomy.html">NexusLIMS Taxonomy</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../user_guide.html" class="nav-link">User Guide</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Extractors</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="extractors">
<span id="id1"></span><h1>Extractors<a class="headerlink" href="#extractors" title="Link to this heading">#</a></h1>
<p>NexusLIMS extracts metadata from various electron microscopy file formats
to create comprehensive experimental records. This page documents the
supported file types, extraction capabilities, and level of support for
each format.</p>
<section id="quick-reference">
<h2>Quick Reference<a class="headerlink" href="#quick-reference" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Instrument/Software</strong></p></th>
<th class="head"><p><strong>Extension</strong></p></th>
<th class="head"><p><strong>Support</strong></p></th>
<th class="head"><p><strong>Data Types</strong></p></th>
<th class="head"><p><strong>Key Features</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#digital-micrograph-files-dm3-dm4">Gatan DigitalMicrograph</a></p></td>
<td><p>.dm3, .dm4</p></td>
<td><p>✅ Full</p></td>
<td><p>TEM/STEM Imaging, EELS, EDS, Diffraction, Spectrum Imaging</p></td>
<td><p>Comprehensive metadata, instrument-specific parsers, automatic type detection</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#feithermo-fisher-tif-files-tif">FEI/Thermo Fisher SEM/FIB</a></p></td>
<td><p>.tif</p></td>
<td><p>✅ Full</p></td>
<td><p>SEM Imaging</p></td>
<td><p>Beam settings, stage position, vacuum conditions, detector config</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#zeiss-orion-fibics-him-tif-files-tif">Zeiss Orion HIM / Fibics HIM</a></p></td>
<td><p>.tif</p></td>
<td><p>✅ Full</p></td>
<td><p>SEM/HIM Imaging</p></td>
<td><p>Helium ion beam settings, stage position, detector configuration, image metadata</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tescan-pfibsem-tif-files-tif">Tescan (P)FIB/SEM</a></p></td>
<td><p>.tif</p></td>
<td><p>✅ Full</p></td>
<td><p>SEM Imaging</p></td>
<td><p>High-voltage settings, stage position, detector gain/offset, scan parameters, stigmator values</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#fei-tia-files-ser-emi">FEI TIA Software</a></p></td>
<td><p>.ser, .emi</p></td>
<td><p>✅ Full</p></td>
<td><p>TEM/STEM Imaging, Diffraction, EELS/EDS Spectra &amp; SI</p></td>
<td><p>Multi-file support, experimental conditions, acquisition parameters</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#edax-eds-files-spc-msa">EDAX (Genesis, TEAM)</a></p></td>
<td><p>.spc</p></td>
<td><p>✅ Full</p></td>
<td><p>EDS Spectrum</p></td>
<td><p>Detector angles, energy calibration, element identification</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#edax-eds-files-spc-msa">EDAX &amp; others (standard)</a></p></td>
<td><p>.msa</p></td>
<td><p>✅ Full</p></td>
<td><p>EDS Spectrum</p></td>
<td><p>EMSA/MAS standard format, vendor extensions supported</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#image-formats">Various (exported images)</a></p></td>
<td><p>.png, .jpg, .tiff, .bmp, .gif</p></td>
<td><p>⚠️ Preview</p></td>
<td><p>Unknown</p></td>
<td><p>Basic metadata, square thumbnail generation</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#text-files-txt">Various (logs, notes)</a></p></td>
<td><p>.txt</p></td>
<td><p>⚠️ Preview</p></td>
<td><p>Unknown</p></td>
<td><p>Basic metadata, text-to-image preview</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#unknown-files">Unknown Files</a></p></td>
<td><p><em>others</em></p></td>
<td><p>❌ Minimal</p></td>
<td><p>Unknown</p></td>
<td><p>Timestamp only, placeholder preview</p></td>
</tr>
</tbody>
</table>
</div>
<p><strong>Legend</strong>: ✅ Full = Comprehensive metadata extraction<br/>⚠️ Preview = Basic metadata + custom preview<br/>❌ Minimal = Timestamp only</p>
</section>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Link to this heading">#</a></h2>
<p>The extraction system uses a <strong>plugin-based architecture</strong> for flexible and extensible metadata extraction. The system consists of three main components:</p>
<ol class="arabic simple">
<li><p><strong>Extractor Plugins</strong>: Parse comprehensive metadata from supported file formats</p></li>
<li><p><strong>Preview Generator Plugins</strong>: Create thumbnail images for visualization</p></li>
<li><p><strong>Extractor Registry</strong>: Manages plugin discovery, selection, and execution</p></li>
</ol>
<section id="plugin-architecture">
<h3>Plugin Architecture<a class="headerlink" href="#plugin-architecture" title="Link to this heading">#</a></h3>
<p>The plugin system provides:</p>
<ul class="simple">
<li><p><strong>Auto-discovery</strong>: Plugins are automatically discovered from <code class="docutils literal notranslate"><span class="pre">nexusLIMS/extractors/plugins/</span></code></p></li>
<li><p><strong>Priority-based selection</strong>: Multiple extractors can support the same file type, with higher priority extractors preferred</p></li>
<li><p><strong>Content sniffing</strong>: Extractors can examine file contents beyond just extensions</p></li>
<li><p><strong>Multi-signal support</strong>: Files containing multiple datasets (e.g., DM3/DM4) are automatically expanded</p></li>
<li><p><strong>Defensive design</strong>: All extractors implement robust error handling with graceful fallbacks</p></li>
</ul>
<p>Extraction is performed automatically during record building. Each file is processed by the best available extractor, and both metadata (saved as JSON) and preview images (saved as PNG thumbnails) are generated in parallel to the original data files.</p>
</section>
</section>
<section id="fully-supported-formats">
<h2>Fully Supported Formats<a class="headerlink" href="#fully-supported-formats" title="Link to this heading">#</a></h2>
<p>These formats have dedicated extractors that parse comprehensive metadata specific to their structure.</p>
<section id="digital-micrograph-files-dm3-dm4">
<span id="id2"></span><h3>Digital Micrograph Files (.dm3, .dm4)<a class="headerlink" href="#digital-micrograph-files-dm3-dm4" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ✅ Full</p>
<p><strong>Description</strong>: Files saved by Gatan’s DigitalMicrograph (GMS) software, commonly used for TEM/STEM imaging, EELS, and EDS data.</p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.digital_micrograph.html#module-nexusLIMS.extractors.plugins.digital_micrograph" title="nexusLIMS.extractors.plugins.digital_micrograph"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.digital_micrograph</span></code></a></p>
<p><strong>Key Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>Microscope information (voltage, magnification, mode, illumination mode)</p></li>
<li><p>Stage position (X, Y, Z, α, β coordinates)</p></li>
<li><p>Acquisition device and camera settings (binning, exposure time)</p></li>
<li><p>Image processing settings</p></li>
<li><p>EELS spectrometer settings (if applicable)</p>
<ul>
<li><p>Acquisition parameters (exposure, integration time, number of frames)</p></li>
<li><p>Experimental conditions (collection/convergence angles)</p></li>
<li><p>Spectrometer configuration (dispersion, energy loss, slit settings)</p></li>
<li><p>Processing information</p></li>
</ul>
</li>
<li><p>EDS detector information (if applicable)</p>
<ul>
<li><p>Acquisition settings (dwell time, dispersion, energy range)</p></li>
<li><p>Detector configuration (angles, window type, solid angle)</p></li>
<li><p>Live/real time and count rates</p></li>
</ul>
</li>
<li><p>Spectrum imaging parameters (if applicable)</p>
<ul>
<li><p>Pixel time, scan mode, spatial sampling</p></li>
<li><p>Drift correction settings</p></li>
<li><p>Acquisition duration</p></li>
</ul>
</li>
</ul>
<p><strong>Instrument-Specific Parsing</strong>:</p>
<p>The extractor includes specialized parsers for specific instruments:</p>
<ul class="simple">
<li><p><strong>FEI Titan STEM</strong> (<code class="docutils literal notranslate"><span class="pre">FEI-Titan-STEM</span></code>): Custom handling for EFTEM diffraction mode detection</p></li>
<li><p><strong>FEI Titan TEM</strong> (<code class="docutils literal notranslate"><span class="pre">FEI-Titan-TEM</span></code>): Parses Tecnai metadata tags including gun settings, lens strengths, apertures, filter settings, and stage positions</p></li>
<li><p><strong>JEOL JEM 3010</strong> (<code class="docutils literal notranslate"><span class="pre">JEOL-JEM-TEM</span></code>): Basic parsing with filename-based diffraction pattern detection</p></li>
</ul>
<p><strong>Data Types Detected</strong>:</p>
<ul class="simple">
<li><p>TEM/STEM Imaging</p></li>
<li><p>TEM/STEM Diffraction</p></li>
<li><p>TEM/STEM EELS (Spectrum)</p></li>
<li><p>TEM/STEM EDS (Spectrum)</p></li>
<li><p>EELS/EDS Spectrum Imaging</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p>Automatically detects dataset type based on metadata (Image, Spectrum, SpectrumImage, Diffraction)</p></li>
<li><p>For stacked images, metadata is extracted from the first plane</p></li>
<li><p>Session info (Operator, Specimen, Detector) may be unreliable and is flagged in warnings</p></li>
</ul>
</section>
<section id="fei-thermo-fisher-tif-files-tif">
<span id="feithermo-fisher-tif-files-tif"></span><h3>FEI/Thermo Fisher TIF Files (.tif)<a class="headerlink" href="#fei-thermo-fisher-tif-files-tif" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ✅ Full</p>
<p><strong>Description</strong>: TIFF images saved by FEI/Thermo Fisher FIB and SEM instruments (Quanta, Helios, etc.) with embedded metadata.</p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.quanta_tif.html#module-nexusLIMS.extractors.plugins.quanta_tif" title="nexusLIMS.extractors.plugins.quanta_tif"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.quanta_tif</span></code></a></p>
<p><strong>Key Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>Beam settings (voltage, emission current, spot size, field widths, working distance)</p></li>
<li><p>Beam positioning (beam shift, tilt, scan rotation)</p></li>
<li><p>Stage position (X, Y, Z, R, α, tilt angles)</p></li>
<li><p>Scan parameters (dwell time, frame time, pixel size, field of view)</p></li>
<li><p>Detector configuration (name, brightness, contrast, signal type, grid voltage)</p></li>
<li><p>System information (software version, chamber type, column type, vacuum pump)</p></li>
<li><p>Vacuum conditions (mode, chamber pressure)</p></li>
<li><p>Image settings (drift correction, frame integration, magnification mode)</p></li>
<li><p>Acquisition date and time</p></li>
<li><p>Specimen temperature (if available)</p></li>
<li><p>User/operator information</p></li>
</ul>
<p><strong>Special Features</strong>:</p>
<ul class="simple">
<li><p>Handles both config-style and XML metadata sections</p></li>
<li><p>Supports MultiGIS gas injection system metadata</p></li>
<li><p>Converts units to display-friendly formats (e.g., SI to μm, μA, etc.)</p></li>
<li><p>Automatic detection and parsing of tilt correction settings</p></li>
</ul>
<p><strong>Data Types Detected</strong>:</p>
<ul class="simple">
<li><p>SEM Imaging</p></li>
</ul>
<p><strong>Preview Generation</strong>:</p>
<ul class="simple">
<li><p>Uses 2× downsampling for efficient thumbnail creation</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p>User/operator metadata is flagged as potentially unreliable (users may remain logged in)</p></li>
<li><p>Some instruments write duplicate metadata sections which are handled automatically</p></li>
<li><p>Works with both older config-style metadata and newer XML-based metadata</p></li>
</ul>
</section>
<section id="zeiss-orion-fibics-him-tif-files-tif">
<span id="id3"></span><h3>Zeiss Orion / Fibics HIM TIF Files (.tif)<a class="headerlink" href="#zeiss-orion-fibics-him-tif-files-tif" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ✅ Full</p>
<p><strong>Description</strong>: TIFF images saved by Zeiss Orion and Fibics helium ion microscope (HIM) systems with embedded XML metadata in custom TIFF tags.</p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.orion_HIM_tif.html#module-nexusLIMS.extractors.plugins.orion_HIM_tif" title="nexusLIMS.extractors.plugins.orion_HIM_tif"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.orion_HIM_tif</span></code></a></p>
<p><strong>File Format Details</strong>:</p>
<p>The extractor uses content sniffing to detect HIM TIFF files by checking for custom TIFF tags:</p>
<ul class="simple">
<li><p><strong>Zeiss Orion</strong>: TIFF tag 65000 contains <code class="docutils literal notranslate"><span class="pre">&lt;ImageTags&gt;</span></code> XML metadata</p></li>
<li><p><strong>Fibics</strong>: TIFF tag 51023 contains <code class="docutils literal notranslate"><span class="pre">&lt;Fibics&gt;</span></code> XML metadata</p></li>
</ul>
<p>This content-based detection allows proper identification even when files use <code class="docutils literal notranslate"><span class="pre">.tif</span></code> extension (used by multiple instruments).</p>
<p><strong>Key Metadata Extracted</strong>:</p>
<p><strong>Zeiss Orion Variant</strong>:</p>
<ul class="simple">
<li><p>Helium ion beam settings (energy, current, spot size, aperture)</p></li>
<li><p>Beam positioning (shift, scan rotation)</p></li>
<li><p>Stage position (X, Y, Z, R, tilt)</p></li>
<li><p>Scan parameters (dwell time, frame time, pixel size, resolution)</p></li>
<li><p>Detector configuration (type, brightness, contrast)</p></li>
<li><p>System information (vacuum pressure, chamber type)</p></li>
<li><p>Image acquisition parameters (line averaging, digital gain, overlay settings)</p></li>
<li><p>Scan speed and resolution settings</p></li>
</ul>
<p><strong>Fibics Variant</strong>:</p>
<ul class="simple">
<li><p>Helium ion beam parameters (energy, current, spot settings)</p></li>
<li><p>Stage coordinates and angles</p></li>
<li><p>Scan configuration (pixel dwell, acquisition time, resolution)</p></li>
<li><p>Detector settings and signal type</p></li>
<li><p>Image optimization parameters (contrast, brightness)</p></li>
<li><p>Data collection timestamps</p></li>
</ul>
<p><strong>Data Types Detected</strong>:</p>
<ul class="simple">
<li><p>HIM Imaging</p></li>
</ul>
<p><strong>Special Features</strong>:</p>
<ul class="simple">
<li><p>Content-based detection to differentiate from other TIFF formats (FEI, standard image files)</p></li>
<li><p>Priority 150 - checked before generic TIFF extractors to properly identify HIM files</p></li>
<li><p>Handles both Zeiss and Fibics XML metadata formats</p></li>
<li><p>Robust error handling for malformed or missing XML metadata</p></li>
<li><p>Supports .tif and .tiff extensions</p></li>
</ul>
<p><strong>Preview Generation</strong>:</p>
<ul class="simple">
<li><p>Converts image to square thumbnail (500×500 px default)</p></li>
<li><p>Maintains aspect ratio with padding</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p>The extractor uses content sniffing rather than extension alone, ensuring correct identification even if .tif files from multiple instruments are present</p></li>
<li><p>If XML metadata is missing or corrupted, the extractor gracefully falls back to basic file information</p></li>
<li><p>Both Zeiss Orion and Fibics HIM variants store metadata as embedded XML, making extraction reliable across different software versions</p></li>
</ul>
</section>
<section id="tescan-pfib-sem-tif-files-tif">
<span id="tescan-pfibsem-tif-files-tif"></span><h3>Tescan PFIB/SEM TIF Files (.tif)<a class="headerlink" href="#tescan-pfib-sem-tif-files-tif" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ✅ Full</p>
<p><strong>Description</strong>: TIFF images saved by Tescan PFIB (Focused Ion Beam) and SEM instruments (e.g., AMBER X) with embedded INI-style metadata in custom TIFF tags or sidecar .hdr files.</p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.tescan_tif.html#module-nexusLIMS.extractors.plugins.tescan_tif" title="nexusLIMS.extractors.plugins.tescan_tif"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.tescan_tif</span></code></a></p>
<p><strong>File Format Details</strong>:</p>
<p>The extractor uses a three-tier strategy for metadata extraction:</p>
<ol class="arabic simple">
<li><p><strong>Primary</strong>: Extracts metadata from embedded TIFF tag 50431 (custom Tescan metadata tag) containing INI-style metadata</p></li>
<li><p><strong>Fallback</strong>: If embedded metadata fails or is incomplete, looks for a sidecar <code class="docutils literal notranslate"><span class="pre">.hdr</span></code> file with full metadata in INI format (<code class="docutils literal notranslate"><span class="pre">[MAIN]</span></code> and <code class="docutils literal notranslate"><span class="pre">[SEM]</span></code> sections)</p></li>
<li><p><strong>Supplementary</strong>: Always extracts basic TIFF tags (tag 271 for Make, tag 305 for Software, tag 315 for Artist) to supplement or override other metadata</p></li>
</ol>
<p>This multi-tier approach ensures complete metadata is available whether metadata is embedded in the TIFF or stored in a sidecar <code class="docutils literal notranslate"><span class="pre">.hdr</span></code> file.</p>
<p><strong>Key Metadata Extracted</strong>:</p>
<p><strong>From [MAIN] section</strong>:</p>
<ul class="simple">
<li><p>Instrument identification (Device, Model, Serial Number)</p></li>
<li><p>User information (Operator name)</p></li>
<li><p>Acquisition timestamp (Date and Time)</p></li>
<li><p>Magnification</p></li>
<li><p>Software version</p></li>
</ul>
<p><strong>From [SEM] section</strong>:</p>
<ul class="simple">
<li><p>Beam parameters (High Voltage, Spot Size, Emission Current)</p></li>
<li><p>Stage position (X, Y, Z coordinates and Rotation/Tilt angles)</p></li>
<li><p>Scan settings (Dwell time, Scan mode, Rotation)</p></li>
<li><p>Detector configuration (Name, Gain, Offset)</p></li>
<li><p>Vacuum conditions (Chamber pressure)</p></li>
<li><p>Stigmator values (X and Y corrections)</p></li>
<li><p>Gun type configuration</p></li>
<li><p>Working Distance</p></li>
<li><p>Session ID for traceability</p></li>
</ul>
<p><strong>Unit Conversions</strong>:</p>
<ul class="simple">
<li><p>Magnification: Converted from raw values to kiloX (kX)</p></li>
<li><p>Voltages: Converted from millivolts to kilovolts (kV)</p></li>
<li><p>Distances: Converted from meters to millimeters (mm) or nanometers (nm) as appropriate</p></li>
<li><p>Currents: Converted from amperes to microamperes (μA)</p></li>
<li><p>Pressure: Converted to millipascals (mPa)</p></li>
<li><p>Pixel sizes: Calculated from image dimensions and field of view, converted to nanometers (nm)</p></li>
</ul>
<p><strong>Data Types Detected</strong>:</p>
<ul class="simple">
<li><p>SEM Imaging</p></li>
</ul>
<p><strong>Special Features</strong>:</p>
<ul class="simple">
<li><p>Priority 150 - Checked before generic TIFF extractors to properly identify Tescan files</p></li>
<li><p>Content-based detection via custom TIFF tags even if <code class="docutils literal notranslate"><span class="pre">.hdr</span></code> file is missing</p></li>
<li><p>Comprehensive stage position tracking (X, Y, Z, Rotation, Tilt)</p></li>
<li><p>Detector settings extraction (Gain and Offset values)</p></li>
<li><p>Automatic conversion of physics units to display-friendly formats</p></li>
<li><p>Empty field exclusion - Fields with empty values are not included in output</p></li>
<li><p>Session tracking with unique Session ID</p></li>
</ul>
<p><strong>Preview Generation</strong>:</p>
<ul class="simple">
<li><p>Converts image to square thumbnail (500×500 px default)</p></li>
<li><p>Maintains aspect ratio with padding</p></li>
</ul>
<p><strong>Warnings</strong>:</p>
<p>The extractor flags the following fields as potentially unreliable:</p>
<ul class="simple">
<li><p><strong>Operator</strong>: May reflect a logged-in user rather than the actual operator who collected the data</p></li>
</ul>
<p><strong>Compatibility Notes</strong>:</p>
<ul class="simple">
<li><p><strong>Tescan AMBER X</strong>: Fully tested and verified</p></li>
<li><p><strong>Other Tescan SEM/PFIB Instruments</strong>: Likely compatible due to consistent INI metadata format, but not yet tested</p></li>
<li><p>Both <code class="docutils literal notranslate"><span class="pre">.tif</span></code> and <code class="docutils literal notranslate"><span class="pre">.tiff</span></code> extensions are supported</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">.hdr</span></code> file is present but cannot be read, the extractor falls back to embedded TIFF tag metadata</p></li>
<li><p>If both sidecar and embedded metadata are available, the sidecar is preferred (more reliable)</p></li>
<li><p>The extractor gracefully handles missing or incomplete metadata sections</p></li>
<li><p>Pixel size is calculated from magnification and field width when not directly available</p></li>
</ul>
</section>
<section id="fei-tia-files-ser-emi">
<span id="id4"></span><h3>FEI TIA Files (.ser, .emi)<a class="headerlink" href="#fei-tia-files-ser-emi" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ✅ Full</p>
<p><strong>Description</strong>: Files saved by FEI’s TIA (Tecnai Imaging and Analysis) software. Data is stored in <code class="docutils literal notranslate"><span class="pre">.ser</span></code> files with accompanying <code class="docutils literal notranslate"><span class="pre">.emi</span></code> metadata files.</p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.fei_emi.html#module-nexusLIMS.extractors.plugins.fei_emi" title="nexusLIMS.extractors.plugins.fei_emi"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.fei_emi</span></code></a></p>
<p><strong>File Relationship</strong>:</p>
<ul class="simple">
<li><p>Each <code class="docutils literal notranslate"><span class="pre">.emi</span></code> file can reference multiple <code class="docutils literal notranslate"><span class="pre">.ser</span></code> data files (named as <code class="docutils literal notranslate"><span class="pre">basename_1.ser</span></code>, <code class="docutils literal notranslate"><span class="pre">basename_2.ser</span></code>, etc.)</p></li>
<li><p>Both files are required for complete metadata extraction</p></li>
<li><p>The extractor automatically locates the corresponding <code class="docutils literal notranslate"><span class="pre">.emi</span></code> file for a given <code class="docutils literal notranslate"><span class="pre">.ser</span></code> file</p></li>
</ul>
<p><strong>Key Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>Manufacturer and acquisition date</p></li>
<li><p>Microscope accelerating voltage and tilt settings</p></li>
<li><p>Acquisition mode and beam position</p></li>
<li><p>Camera settings (name, binning, dwell time, frame time)</p></li>
<li><p>Detector configuration (energy resolution, integration time)</p></li>
<li><p>Scan parameters (area, drift correction, spectra count)</p></li>
<li><p>Experimental conditions from TIA software</p></li>
</ul>
<p><strong>Data Types Detected</strong>:</p>
<ul class="simple">
<li><p>TEM/STEM Imaging</p></li>
<li><p>TEM/STEM Diffraction</p></li>
<li><p>EELS/EDS Spectrum and Spectrum Imaging</p></li>
</ul>
<p><strong>Type Detection Logic</strong>:</p>
<ul class="simple">
<li><p>Uses <code class="docutils literal notranslate"><span class="pre">Mode</span></code> metadata field (if present) to distinguish TEM/STEM and Image/Diffraction</p></li>
<li><p>Signal dimension determines Image vs. Spectrum</p></li>
<li><p>Navigation dimension presence indicates Spectrum Imaging</p></li>
<li><p>Heuristic analysis of axis values used to distinguish EELS vs. EDS when not explicitly labeled</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">.emi</span></code> file is missing, extractor falls back to <code class="docutils literal notranslate"><span class="pre">.ser</span></code> file only (limited metadata)</p></li>
<li><p>Multiple signals in one <code class="docutils literal notranslate"><span class="pre">.emi</span></code> file are handled; metadata is extracted from the appropriate index</p></li>
<li><p>Later signals in a multi-file series may have less metadata than the first</p></li>
</ul>
</section>
<section id="edax-eds-files-spc-msa">
<span id="id5"></span><h3>EDAX EDS Files (.spc, .msa)<a class="headerlink" href="#edax-eds-files-spc-msa" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ✅ Full</p>
<p><strong>Description</strong>: EDS spectrum files saved by EDAX software (Genesis, TEAM, etc.) in proprietary (<code class="docutils literal notranslate"><span class="pre">.spc</span></code>) or standard EMSA (<code class="docutils literal notranslate"><span class="pre">.msa</span></code>) format.</p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.edax.html#module-nexusLIMS.extractors.plugins.edax" title="nexusLIMS.extractors.plugins.edax"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.edax</span></code></a></p>
<section id="spc-files">
<h4>.spc Files<a class="headerlink" href="#spc-files" title="Link to this heading">#</a></h4>
<p><strong>Key Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>Azimuthal and elevation angles</p></li>
<li><p>Live time</p></li>
<li><p>Detector energy resolution</p></li>
<li><p>Accelerating voltage</p></li>
<li><p>Channel size and energy range</p></li>
<li><p>Number of spectrum channels</p></li>
<li><p>Stage tilt</p></li>
<li><p>Identified elements</p></li>
</ul>
<p><strong>Data Types Detected</strong>:</p>
<ul class="simple">
<li><p>EDS Spectrum</p></li>
</ul>
</section>
<section id="msa-files">
<h4>.msa Files<a class="headerlink" href="#msa-files" title="Link to this heading">#</a></h4>
<p><strong>Description</strong>: MSA (EMSA/MAS) format is a standard spectral data format. See the <a class="reference external" href="https://www.microscopy.org/resources/scientific_data/">Microscopy Society of America specification</a>.</p>
<p><strong>Key Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>All standard MSA fields (version, format, data dimensions)</p></li>
<li><p>EDAX-specific extensions (angles, times, resolutions)</p></li>
<li><p>Analyzer and detector configuration</p></li>
<li><p>User-selected elements</p></li>
<li><p>Amplifier settings</p></li>
<li><p>FPGA version</p></li>
<li><p>Originating file information</p></li>
<li><p>Comments and title</p></li>
</ul>
<p><strong>Data Types Detected</strong>:</p>
<ul class="simple">
<li><p>EDS Spectrum</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">.msa</span></code> files are vendor-agnostic and may be exported from various EDS software</p></li>
<li><p>EDAX adds custom fields beyond the MSA standard</p></li>
<li><p>Both formats are single-spectrum only (not spectrum images)</p></li>
</ul>
</section>
</section>
</section>
<section id="partially-supported-formats">
<h2>Partially Supported Formats<a class="headerlink" href="#partially-supported-formats" title="Link to this heading">#</a></h2>
<p>These formats receive basic metadata extraction and custom preview generation, but do not have dedicated metadata parsers.</p>
<section id="image-formats">
<span id="id6"></span><h3>Image Formats<a class="headerlink" href="#image-formats" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ⚠️ Preview Only</p>
<p><strong>Formats</strong>: <code class="docutils literal notranslate"><span class="pre">.png</span></code>, <code class="docutils literal notranslate"><span class="pre">.tiff</span></code>, <code class="docutils literal notranslate"><span class="pre">.bmp</span></code>, <code class="docutils literal notranslate"><span class="pre">.gif</span></code>, <code class="docutils literal notranslate"><span class="pre">.jpg</span></code>, <code class="docutils literal notranslate"><span class="pre">.jpeg</span></code></p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.basic_metadata.html#module-nexusLIMS.extractors.plugins.basic_metadata" title="nexusLIMS.extractors.plugins.basic_metadata"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.basic_metadata</span></code></a></p>
<p><strong>Preview Generator</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.preview_generators.image_preview.html#module-nexusLIMS.extractors.plugins.preview_generators.image_preview" title="nexusLIMS.extractors.plugins.preview_generators.image_preview"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.preview_generators.image_preview</span></code></a></p>
<p><strong>Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>File creation/modification time</p></li>
<li><p>Instrument ID (inferred from file path)</p></li>
</ul>
<p><strong>Preview Generation</strong>:</p>
<ul class="simple">
<li><p>Converts image to square thumbnail (500×500 px default)</p></li>
<li><p>Maintains aspect ratio with padding</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p>These are typically auxiliary files (screenshots, exported images, etc.)</p></li>
<li><p>Marked as <code class="docutils literal notranslate"><span class="pre">DatasetType:</span> <span class="pre">Unknown</span></code> in records</p></li>
</ul>
</section>
<section id="text-files-txt">
<span id="id7"></span><h3>Text Files (.txt)<a class="headerlink" href="#text-files-txt" title="Link to this heading">#</a></h3>
<p><strong>Support Level</strong>: ⚠️ Preview Only</p>
<p><strong>Extractor Module</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.basic_metadata.html#module-nexusLIMS.extractors.plugins.basic_metadata" title="nexusLIMS.extractors.plugins.basic_metadata"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.basic_metadata</span></code></a></p>
<p><strong>Preview Generator</strong>: <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.preview_generators.text_preview.html#module-nexusLIMS.extractors.plugins.preview_generators.text_preview" title="nexusLIMS.extractors.plugins.preview_generators.text_preview"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.preview_generators.text_preview</span></code></a></p>
<p><strong>Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>File creation/modification time</p></li>
<li><p>Instrument ID (inferred from file path)</p></li>
</ul>
<p><strong>Preview Generation</strong>:</p>
<ul class="simple">
<li><p>Renders first ~20 lines of text as image thumbnail</p></li>
<li><p>Uses monospace font for readability</p></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul class="simple">
<li><p>Common for log files, notes, and exported data</p></li>
<li><p>Marked as <code class="docutils literal notranslate"><span class="pre">DatasetType:</span> <span class="pre">Unknown</span></code> in records</p></li>
</ul>
</section>
</section>
<section id="unsupported-formats">
<span id="unknown-files"></span><h2>Unsupported Formats<a class="headerlink" href="#unsupported-formats" title="Link to this heading">#</a></h2>
<p><strong>Support Level</strong>: ❌ Minimal</p>
<p>Files with extensions not in the above lists receive minimal processing:</p>
<p><strong>Metadata Extracted</strong>:</p>
<ul class="simple">
<li><p>File creation/modification time only</p></li>
<li><p>Marked as <code class="docutils literal notranslate"><span class="pre">DatasetType:</span> <span class="pre">Unknown</span></code> and <code class="docutils literal notranslate"><span class="pre">Data</span> <span class="pre">Type:</span> <span class="pre">Unknown</span></code></p></li>
</ul>
<p><strong>Preview Generation</strong>:</p>
<ul class="simple">
<li><p>A placeholder image is used indicating extraction failed</p></li>
</ul>
<p><strong>Handling Strategy</strong>:</p>
<p>The system’s behavior for unsupported files depends on the <code class="docutils literal notranslate"><span class="pre">NX_FILE_STRATEGY</span></code> environment variable:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">exclusive</span></code> (default): Only files with full extractors are included in records</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">inclusive</span></code>: All files are included, with basic metadata for unsupported types</p></li>
</ul>
</section>
<section id="how-extraction-works">
<h2>How Extraction Works<a class="headerlink" href="#how-extraction-works" title="Link to this heading">#</a></h2>
<section id="file-discovery-and-strategy">
<h3>File Discovery and Strategy<a class="headerlink" href="#file-discovery-and-strategy" title="Link to this heading">#</a></h3>
<p>During record building, NexusLIMS finds files within the session time window using the configured strategy:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Only include files with dedicated extractors</span>
<span class="nv">NX_FILE_STRATEGY</span><span class="o">=</span>exclusive

<span class="c1"># Include all files found</span>
<span class="nv">NX_FILE_STRATEGY</span><span class="o">=</span>inclusive
</pre></div>
</div>
</section>
<section id="extraction-process">
<h3>Extraction Process<a class="headerlink" href="#extraction-process" title="Link to this heading">#</a></h3>
<p>For each discovered file:</p>
<ol class="arabic simple">
<li><p><strong>Plugin Discovery</strong>: The extractor registry auto-discovers all available extractor plugins from <code class="docutils literal notranslate"><span class="pre">nexusLIMS/extractors/plugins/</span></code></p></li>
<li><p><strong>Extractor Selection</strong>:</p>
<ul class="simple">
<li><p>The registry uses priority-based selection with content sniffing support</p></li>
<li><p>Extractors are tried in descending priority order (higher priority first)</p></li>
<li><p>Each extractor’s <code class="docutils literal notranslate"><span class="pre">supports()</span></code> method is called to determine compatibility</p></li>
<li><p>If no specialized extractor matches, the <code class="docutils literal notranslate"><span class="pre">BasicFileInfoExtractor</span></code> fallback is used</p></li>
</ul>
</li>
<li><p><strong>Metadata Parsing</strong>: The selected extractor reads the file and returns a <strong>list</strong> of metadata dictionaries:</p>
<ul class="simple">
<li><p>Each dictionary contains <code class="docutils literal notranslate"><span class="pre">nx_meta</span></code> with NexusLIMS-specific metadata (standardized keys)</p></li>
<li><p>Additional keys contain format-specific “raw” metadata</p></li>
<li><p>For multi-signal files, the list contains one entry per signal</p></li>
<li><p>For single-signal files, the list contains one entry for consistency</p></li>
</ul>
</li>
<li><p><strong>Metadata Writing</strong>: JSON file(s) are written to parallel path in <code class="docutils literal notranslate"><span class="pre">NX_DATA_PATH</span></code></p>
<ul class="simple">
<li><p>Path: <code class="docutils literal notranslate"><span class="pre">{NX_DATA_PATH}/{instrument}/{path/to/file}.json</span></code></p></li>
<li><p>For multi-signal files: Multiple JSON files with <code class="docutils literal notranslate"><span class="pre">_signalN</span></code> suffixes</p></li>
</ul>
</li>
<li><p><strong>Preview Generation</strong>: Thumbnail PNG(s) are created</p>
<ul class="simple">
<li><p>Path: <code class="docutils literal notranslate"><span class="pre">{NX_DATA_PATH}/{instrument}/{path/to/file}.thumb.png</span></code></p></li>
<li><p>For multi-signal files: Multiple previews with <code class="docutils literal notranslate"><span class="pre">_signalN.thumb.png</span></code> suffixes</p></li>
<li><p>Size: 500×500 px (default)</p></li>
<li><p>Uses plugin-based preview generators with fallback to legacy methods</p></li>
</ul>
</li>
</ol>
</section>
<section id="expected-metadata-structure">
<h3>Expected Metadata Structure<a class="headerlink" href="#expected-metadata-structure" title="Link to this heading">#</a></h3>
<p>All extractors return a <strong>list of metadata dictionaries</strong>, with one entry per signal or dataset:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">[</span>
    <span class="p">{</span>
        <span class="s2">&quot;nx_meta&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Creation Time&quot;</span><span class="p">:</span> <span class="s2">&quot;ISO 8601 timestamp with timezone&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Data Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Category_Modality_Technique&quot;</span><span class="p">,</span>  <span class="c1"># e.g., &quot;STEM_EDS&quot;</span>
            <span class="s2">&quot;DatasetType&quot;</span><span class="p">:</span> <span class="s2">&quot;Image|Spectrum|SpectrumImage|Diffraction|Misc|Unknown&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Data Dimensions&quot;</span><span class="p">:</span> <span class="s2">&quot;(height, width)&quot;</span> <span class="ow">or</span> <span class="s2">&quot;(channels,)&quot;</span><span class="p">,</span>  <span class="c1"># Optional</span>
            <span class="s2">&quot;Instrument ID&quot;</span><span class="p">:</span> <span class="s2">&quot;instrument-name&quot;</span><span class="p">,</span>  <span class="c1"># Optional</span>
            <span class="s2">&quot;warnings&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;field1&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;field2&quot;</span><span class="p">]],</span>  <span class="c1"># Optional, list of field names flagged as unreliable</span>
            <span class="c1"># ... format-specific keys ...</span>
        <span class="p">},</span>
        <span class="c1"># Additional format-specific metadata sections</span>
        <span class="s2">&quot;ImageList&quot;</span><span class="p">:</span> <span class="p">{</span> <span class="o">...</span> <span class="p">},</span>  <span class="c1"># Example: DM3/DM4 files</span>
        <span class="s2">&quot;ObjectInfo&quot;</span><span class="p">:</span> <span class="p">{</span> <span class="o">...</span> <span class="p">},</span>  <span class="c1"># Example: FEI .ser/.emi files</span>
        <span class="c1"># etc.</span>
    <span class="p">},</span>
    <span class="c1"># For multi-signal files, additional entries follow the same structure</span>
    <span class="p">{</span>
        <span class="s2">&quot;nx_meta&quot;</span><span class="p">:</span> <span class="p">{</span> <span class="o">...</span> <span class="p">},</span>
        <span class="c1"># ... additional signal ...</span>
    <span class="p">}</span>
<span class="p">]</span>
</pre></div>
</div>
<p><strong>Return Format:</strong></p>
<ul class="simple">
<li><p><strong>Single-signal files</strong>: Return a list with one element <code class="docutils literal notranslate"><span class="pre">[{...}]</span></code></p></li>
<li><p><strong>Multi-signal files</strong>: Return a list with multiple elements, one per signal <code class="docutils literal notranslate"><span class="pre">[{...},</span> <span class="pre">{...}]</span></code></p></li>
</ul>
<p><strong>Validation Using Pydantic Models (New in v2.2.0 - Work in Progress):</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">nx_meta</span></code> section is validated against Pydantic schemas to improve data consistency and quality. Starting with v2.2.0, NexusLIMS provides basic metadata validation with EM Glossary integration and unit standardization. This validation system is still evolving as standardization efforts progress.</p>
<p><strong>Schema Selection by DatasetType:</strong></p>
<p>The validation schema is chosen automatically based on the <code class="docutils literal notranslate"><span class="pre">DatasetType</span></code> field:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>DatasetType</strong></p></th>
<th class="head"><p><strong>Schema</strong></p></th>
<th class="head"><p><strong>Purpose</strong></p></th>
<th class="head"><p><strong>Type-Specific Fields</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">Image</span></code></p></td>
<td><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.schemas.metadata.html#nexusLIMS.schemas.metadata.ImageMetadata" title="nexusLIMS.schemas.metadata.ImageMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">ImageMetadata</span></code></a></p></td>
<td><p>TEM/SEM/STEM images</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">magnification</span></code>, <code class="docutils literal notranslate"><span class="pre">pixel_size</span></code>, <code class="docutils literal notranslate"><span class="pre">working_distance</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">Spectrum</span></code></p></td>
<td><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.schemas.metadata.html#nexusLIMS.schemas.metadata.SpectrumMetadata" title="nexusLIMS.schemas.metadata.SpectrumMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">SpectrumMetadata</span></code></a></p></td>
<td><p>EELS/EDS spectra</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">channel_size</span></code>, <code class="docutils literal notranslate"><span class="pre">starting_energy</span></code>, <code class="docutils literal notranslate"><span class="pre">live_time</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">SpectrumImage</span></code></p></td>
<td><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.schemas.metadata.html#nexusLIMS.schemas.metadata.SpectrumImageMetadata" title="nexusLIMS.schemas.metadata.SpectrumImageMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">SpectrumImageMetadata</span></code></a></p></td>
<td><p>Spectrum imaging</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">pixel_time</span></code>, <code class="docutils literal notranslate"><span class="pre">magnification</span></code>, <code class="docutils literal notranslate"><span class="pre">channel_size</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">Diffraction</span></code></p></td>
<td><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.schemas.metadata.html#nexusLIMS.schemas.metadata.DiffractionMetadata" title="nexusLIMS.schemas.metadata.DiffractionMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiffractionMetadata</span></code></a></p></td>
<td><p>Diffraction patterns</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">camera_length</span></code>, <code class="docutils literal notranslate"><span class="pre">exposure_time</span></code>, <code class="docutils literal notranslate"><span class="pre">convergence_angle</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">Misc</span></code> or <code class="docutils literal notranslate"><span class="pre">Unknown</span></code></p></td>
<td><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.schemas.metadata.html#nexusLIMS.schemas.metadata.NexusMetadata" title="nexusLIMS.schemas.metadata.NexusMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">NexusMetadata</span></code></a></p></td>
<td><p>Other data</p></td>
<td><p>Base fields only</p></td>
</tr>
</tbody>
</table>
</div>
<p><strong>Common Core Fields (All Schemas):</strong></p>
<p>All schemas inherit these required fields from <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.schemas.metadata.html#nexusLIMS.schemas.metadata.NexusMetadata" title="nexusLIMS.schemas.metadata.NexusMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">nexusLIMS.schemas.metadata.NexusMetadata</span></code></a>:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">creation_time</span></code></strong> (datetime with timezone): When the data was acquired</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">data_type</span></code></strong> (str): Classification string (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;STEM_Imaging&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;TEM_EDS&quot;</span></code>)</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">dataset_type</span></code></strong> (DatasetType enum): One of <code class="docutils literal notranslate"><span class="pre">Image</span></code>, <code class="docutils literal notranslate"><span class="pre">Spectrum</span></code>, <code class="docutils literal notranslate"><span class="pre">SpectrumImage</span></code>, <code class="docutils literal notranslate"><span class="pre">Diffraction</span></code>, <code class="docutils literal notranslate"><span class="pre">Misc</span></code>, <code class="docutils literal notranslate"><span class="pre">Unknown</span></code></p></li>
</ul>
<p><strong>Optional Core Fields:</strong></p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">data_dimensions</span></code></strong> (str): Shape of the data (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;(1024,</span> <span class="pre">1024)&quot;</span></code>)</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">instrument_id</span></code></strong> (str): Identifier of the instrument used</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">warnings</span></code></strong> (list[str]): Field names flagged as potentially unreliable</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">extensions</span></code></strong> (dict): Additional vendor/instrument-specific metadata</p></li>
</ul>
<p><strong>EM Glossary Standardized Fields:</strong></p>
<p>Type-specific schemas include fields standardized against the Electron Microscopy Glossary v2.0 where available. Note that the EM Glossary vocabulary is also still developing, so not all fields have complete standardized definitions yet:</p>
<ul class="simple">
<li><p><strong>Beam parameters</strong>: <code class="docutils literal notranslate"><span class="pre">acceleration_voltage</span></code>, <code class="docutils literal notranslate"><span class="pre">beam_current</span></code>, <code class="docutils literal notranslate"><span class="pre">emission_current</span></code>, <code class="docutils literal notranslate"><span class="pre">convergence_angle</span></code></p></li>
<li><p><strong>Stage/Sample</strong>: <code class="docutils literal notranslate"><span class="pre">stage_x</span></code>, <code class="docutils literal notranslate"><span class="pre">stage_y</span></code>, <code class="docutils literal notranslate"><span class="pre">stage_z</span></code>, <code class="docutils literal notranslate"><span class="pre">tilt_alpha</span></code>, <code class="docutils literal notranslate"><span class="pre">tilt_beta</span></code></p></li>
<li><p><strong>Detector</strong>: <code class="docutils literal notranslate"><span class="pre">detector_type</span></code>, <code class="docutils literal notranslate"><span class="pre">detector_angle</span></code>, <code class="docutils literal notranslate"><span class="pre">detector_distance</span></code>, <code class="docutils literal notranslate"><span class="pre">detector_binning</span></code></p></li>
<li><p><strong>Acquisition</strong>: <code class="docutils literal notranslate"><span class="pre">dwell_time</span></code>, <code class="docutils literal notranslate"><span class="pre">acquisition_time</span></code>, <code class="docutils literal notranslate"><span class="pre">live_time</span></code>, <code class="docutils literal notranslate"><span class="pre">pixel_time</span></code></p></li>
<li><p><strong>Optical</strong>: <code class="docutils literal notranslate"><span class="pre">magnification</span></code>, <code class="docutils literal notranslate"><span class="pre">working_distance</span></code>, <code class="docutils literal notranslate"><span class="pre">camera_length</span></code>, <code class="docutils literal notranslate"><span class="pre">pixel_size_x</span></code>, <code class="docutils literal notranslate"><span class="pre">pixel_size_y</span></code></p></li>
<li><p><strong>Spectrum</strong>: <code class="docutils literal notranslate"><span class="pre">channel_size</span></code>, <code class="docutils literal notranslate"><span class="pre">starting_energy</span></code>, <code class="docutils literal notranslate"><span class="pre">takeoff_angle</span></code></p></li>
</ul>
<p>These fields are defined using the <code class="docutils literal notranslate"><span class="pre">emg_field()</span></code> helper function, which automatically injects EM Glossary metadata (display names, definitions, URIs) for standardization and interoperability where EMG terms are available.</p>
<p>See <a class="reference internal" href="../dev_guide/em_glossary_reference.html"><span class="doc">EM Glossary Field Reference</span></a> for the complete field mapping and <a class="reference internal" href="../dev_guide/nexuslims_internal_schema.html"><span class="doc">Internal Metadata Schema System</span></a> for details on the schema system.</p>
<p><strong>Validation Example:</strong></p>
<p>When an extractor returns metadata, NexusLIMS automatically validates it:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">datetime</span><span class="w"> </span><span class="kn">import</span> <span class="n">datetime</span><span class="p">,</span> <span class="n">timezone</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">nexusLIMS.schemas.metadata</span><span class="w"> </span><span class="kn">import</span> <span class="n">ImageMetadata</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pint</span><span class="w"> </span><span class="kn">import</span> <span class="n">Quantity</span>

<span class="c1"># Valid ImageMetadata that passes validation</span>
<span class="n">nx_meta</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;creation_time&quot;</span><span class="p">:</span> <span class="n">datetime</span><span class="p">(</span><span class="mi">2024</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">tzinfo</span><span class="o">=</span><span class="n">timezone</span><span class="o">.</span><span class="n">utc</span><span class="p">),</span>
    <span class="s2">&quot;data_type&quot;</span><span class="p">:</span> <span class="s2">&quot;STEM_Imaging&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dataset_type&quot;</span><span class="p">:</span> <span class="s2">&quot;Image&quot;</span><span class="p">,</span>
    <span class="s2">&quot;data_dimensions&quot;</span><span class="p">:</span> <span class="s2">&quot;(1024, 1024)&quot;</span><span class="p">,</span>
    <span class="s2">&quot;instrument_id&quot;</span><span class="p">:</span> <span class="s2">&quot;FEI-Titan-TEM-635816&quot;</span><span class="p">,</span>
    <span class="s2">&quot;warnings&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Operator&quot;</span><span class="p">],</span>  <span class="c1"># Flag unreliable fields</span>
    <span class="c1"># EM Glossary standardized fields with Pint Quantities</span>
    <span class="s2">&quot;acceleration_voltage&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="s2">&quot;kV&quot;</span><span class="p">),</span>
    <span class="s2">&quot;magnification&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">50000</span><span class="p">,</span> <span class="s2">&quot;dimensionless&quot;</span><span class="p">),</span>
    <span class="s2">&quot;pixel_size_x&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mf">2.5</span><span class="p">,</span> <span class="s2">&quot;nm&quot;</span><span class="p">),</span>
    <span class="s2">&quot;working_distance&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mf">5.2</span><span class="p">,</span> <span class="s2">&quot;mm&quot;</span><span class="p">),</span>
    <span class="c1"># Extensions for vendor-specific or non-standardized data</span>
    <span class="s2">&quot;extensions&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;detector_brightness&quot;</span><span class="p">:</span> <span class="mf">50.0</span><span class="p">,</span>
        <span class="s2">&quot;scan_rotation&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">45</span><span class="p">,</span> <span class="s2">&quot;degree&quot;</span><span class="p">),</span>
    <span class="p">},</span>
<span class="p">}</span>
<span class="n">validated</span> <span class="o">=</span> <span class="n">ImageMetadata</span><span class="o">.</span><span class="n">model_validate</span><span class="p">(</span><span class="n">nx_meta</span><span class="p">)</span>

<span class="c1"># Invalid metadata raises ValidationError with details</span>
<span class="n">bad_meta</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;creation_time&quot;</span><span class="p">:</span> <span class="n">datetime</span><span class="p">(</span><span class="mi">2024</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>  <span class="c1"># ❌ Missing timezone!</span>
    <span class="s2">&quot;data_type&quot;</span><span class="p">:</span> <span class="s2">&quot;STEM_Imaging&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dataset_type&quot;</span><span class="p">:</span> <span class="s2">&quot;InvalidType&quot;</span><span class="p">,</span>  <span class="c1"># ❌ Not in allowed values!</span>
<span class="p">}</span>
<span class="n">ImageMetadata</span><span class="o">.</span><span class="n">model_validate</span><span class="p">(</span><span class="n">bad_meta</span><span class="p">)</span>  <span class="c1"># Raises pydantic.ValidationError</span>
</pre></div>
</div>
<p><strong>Unit Handling:</strong></p>
<p>NexusLIMS automatically normalizes physical quantities to preferred units:</p>
<ul class="simple">
<li><p>Voltages → kilovolts (kV)</p></li>
<li><p>Distances → nanometers (nm) or millimeters (mm) depending on scale</p></li>
<li><p>Times → seconds (s)</p></li>
<li><p>Energies → electron volts (eV)</p></li>
</ul>
<p>You can provide quantities in any compatible unit, and validation will convert them automatically:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># These are all equivalent after validation:</span>
<span class="p">{</span><span class="s2">&quot;acceleration_voltage&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="s2">&quot;kV&quot;</span><span class="p">)}</span>
<span class="p">{</span><span class="s2">&quot;acceleration_voltage&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">200000</span><span class="p">,</span> <span class="s2">&quot;V&quot;</span><span class="p">)}</span>
<span class="p">{</span><span class="s2">&quot;acceleration_voltage&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="s2">&quot;MV&quot;</span><span class="p">)}</span>
</pre></div>
</div>
<p>See <a class="reference internal" href="../dev_guide/nexuslims_internal_schema.html"><span class="doc">Internal Metadata Schema System</span></a> for complete unit handling details.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">nx_meta</span></code> section in each element contains standardized, human-readable metadata that is displayed in the experimental record. The additional sections contain the complete “raw” metadata tree for reference.</p>
<p>This consistent list-based approach combined with Pydantic validation ensures the Activity layer can automatically and safely expand multi-signal files into multiple datasets in the experimental record.</p>
</section>
<section id="extension-fields-in-practice">
<h3>Extension Fields in Practice<a class="headerlink" href="#extension-fields-in-practice" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">extensions</span></code> dictionary provides a flexible way to include vendor-specific or non-standardized metadata that doesn’t fit into the core schema fields. Here are real-world examples from NexusLIMS extractors:</p>
<p><strong>FEI/Thermo Quanta TIF Extractor</strong> (<a class="reference download internal" download="" href="../_downloads/9aab50f0d31c81483ce18a2ba7c19905/quanta_tif.py"><span class="xref download myst">nexusLIMS/extractors/plugins/quanta_tif.py:428</span></a>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Vendor-specific vacuum and detector settings</span>
<span class="n">extensions</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;chamber_type&quot;</span><span class="p">:</span> <span class="s2">&quot;LargeChamber&quot;</span><span class="p">,</span>
    <span class="s2">&quot;column_type&quot;</span><span class="p">:</span> <span class="s2">&quot;HighResolution&quot;</span><span class="p">,</span>
    <span class="s2">&quot;vacuum_mode&quot;</span><span class="p">:</span> <span class="s2">&quot;HighVacuum&quot;</span><span class="p">,</span>
    <span class="s2">&quot;chamber_pressure&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mf">5.2e-5</span><span class="p">,</span> <span class="s2">&quot;Pa&quot;</span><span class="p">),</span>
    <span class="s2">&quot;detector_signal&quot;</span><span class="p">:</span> <span class="s2">&quot;ETD&quot;</span><span class="p">,</span>
    <span class="s2">&quot;detector_grid_voltage&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="s2">&quot;V&quot;</span><span class="p">),</span>
    <span class="s2">&quot;drift_correction_enabled&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s2">&quot;frame_integration&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
<p><strong>Digital Micrograph DM3/DM4 Extractor</strong> (<a class="reference download internal" download="" href="../_downloads/357eee9fd8c02bcf4994819d7b8faec8/digital_micrograph.py"><span class="xref download myst">nexusLIMS/extractors/plugins/digital_micrograph.py:337</span></a>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># EELS-specific detector and acquisition settings</span>
<span class="n">extensions</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;eels_dispersion&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;eV/channel&quot;</span><span class="p">),</span>
    <span class="s2">&quot;eels_energy_loss&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">150</span><span class="p">,</span> <span class="s2">&quot;eV&quot;</span><span class="p">),</span>
    <span class="s2">&quot;eels_slit_width&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mf">2.5</span><span class="p">,</span> <span class="s2">&quot;eV&quot;</span><span class="p">),</span>
    <span class="s2">&quot;eels_collection_angle&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="s2">&quot;mrad&quot;</span><span class="p">),</span>
    <span class="s2">&quot;spectrometer_name&quot;</span><span class="p">:</span> <span class="s2">&quot;GIF Quantum 965&quot;</span><span class="p">,</span>
    <span class="s2">&quot;drift_correction_enabled&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
<p><strong>Tescan TIF Extractor</strong> (<a class="reference download internal" download="" href="../_downloads/0532b399b6fc234742d192cd1fef94ae/tescan_tif.py"><span class="xref download myst">nexusLIMS/extractors/plugins/tescan_tif.py:285</span></a>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Instrument-specific settings not covered by EM Glossary</span>
<span class="n">extensions</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;stigmator_x&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="mf">12.5</span><span class="p">,</span> <span class="s2">&quot;percent&quot;</span><span class="p">),</span>
    <span class="s2">&quot;stigmator_y&quot;</span><span class="p">:</span> <span class="n">Quantity</span><span class="p">(</span><span class="o">-</span><span class="mf">8.3</span><span class="p">,</span> <span class="s2">&quot;percent&quot;</span><span class="p">),</span>
    <span class="s2">&quot;gun_type&quot;</span><span class="p">:</span> <span class="s2">&quot;Schottky&quot;</span><span class="p">,</span>
    <span class="s2">&quot;scan_mode&quot;</span><span class="p">:</span> <span class="s2">&quot;FrameIntegration&quot;</span><span class="p">,</span>
    <span class="s2">&quot;session_id&quot;</span><span class="p">:</span> <span class="s2">&quot;abc123xyz789&quot;</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
<p><strong>When to Use Extensions:</strong></p>
<p>Use the <code class="docutils literal notranslate"><span class="pre">extensions</span></code> dictionary when:</p>
<ul class="simple">
<li><p><strong>Vendor-specific features</strong>: Settings unique to one manufacturer (e.g., FEI MultiGIS, Gatan GIF)</p></li>
<li><p><strong>Non-standardized parameters</strong>: Values not yet in EM Glossary (e.g., drift correction flags)</p></li>
<li><p><strong>Instrument-specific calibrations</strong>: Stigmator values, lens strengths, custom aperture settings</p></li>
<li><p><strong>Session tracking</strong>: Unique IDs or timestamps for instrument logging</p></li>
<li><p><strong>Software-specific metadata</strong>: Settings from acquisition software that don’t map to standard fields</p></li>
</ul>
<p><strong>Do NOT use extensions for:</strong></p>
<ul class="simple">
<li><p>Standard EM parameters covered by EM Glossary (voltage, magnification, stage position, etc.)</p></li>
<li><p>Required fields (creation_time, data_type, dataset_type)</p></li>
<li><p>Data that should be validated (use core fields for type safety)</p></li>
</ul>
<p>See <a class="reference internal" href="../dev_guide/nexuslims_internal_schema.html#metadata-helper-functions"><span class="std std-ref">Helper Functions</span></a> for the <code class="docutils literal notranslate"><span class="pre">add_to_extensions()</span></code> helper function and <a class="reference internal" href="../dev_guide/writing_extractor_plugins.html"><span class="doc">Writing Extractor Plugins</span></a> for complete guidance on using extensions in your own extractors.</p>
</section>
</section>
<section id="adding-support-for-new-formats">
<h2>Adding Support for New Formats<a class="headerlink" href="#adding-support-for-new-formats" title="Link to this heading">#</a></h2>
<p>See <a class="reference internal" href="../dev_guide/writing_extractor_plugins.html"><span class="doc">Writing Extractor Plugins</span></a> for instructions on how to write a new extractor.</p>
</section>
<section id="api-reference">
<h2>API Reference<a class="headerlink" href="#api-reference" title="Link to this heading">#</a></h2>
<section id="extractor-registry-properties">
<h3>Extractor Registry Properties<a class="headerlink" href="#extractor-registry-properties" title="Link to this heading">#</a></h3>
<p>The <a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.registry.html#nexusLIMS.extractors.registry.ExtractorRegistry" title="nexusLIMS.extractors.registry.ExtractorRegistry"><code class="xref py py-class docutils literal notranslate"><span class="pre">nexusLIMS.extractors.registry.ExtractorRegistry</span></code></a> class provides convenient properties for querying registered extractors:</p>
<p><strong><code class="docutils literal notranslate"><span class="pre">extractors</span></code> Property</strong>
: Returns a dictionary mapping file extensions to lists of extractor classes, sorted by priority (descending). This property automatically triggers plugin discovery if not already performed.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">nexusLIMS.extractors.registry</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_registry</span>

<span class="n">registry</span> <span class="o">=</span> <span class="n">get_registry</span><span class="p">()</span>
<span class="n">extractors_by_ext</span> <span class="o">=</span> <span class="n">registry</span><span class="o">.</span><span class="n">extractors</span>
<span class="c1"># Returns: {</span>
<span class="c1">#   &#39;dm3&#39;: [&lt;class digital_micrograph.DM3Extractor&#39;&gt;], </span>
<span class="c1">#   &#39;dm4&#39;: [&lt;class &#39;digital_micrograph.DM3Extractor&#39;&gt;], </span>
<span class="c1">#   &#39;msa&#39;: [&lt;class &#39;edax.MsaExtractor&#39;&gt;], </span>
<span class="c1">#   &#39;spc&#39;: [&lt;class &#39;edax.SpcExtractor&#39;&gt;], </span>
<span class="c1">#   ... </span>
<span class="c1"># }</span>
</pre></div>
</div>
<p><strong><code class="docutils literal notranslate"><span class="pre">extractor_names</span></code> Property</strong>
: Returns a deduplicated, alphabetically-sorted list of all registered extractor class names. Includes both extension-specific and wildcard extractors. This property also triggers auto-discovery if needed.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">registry</span> <span class="o">=</span> <span class="n">get_registry</span><span class="p">()</span>
<span class="n">names</span> <span class="o">=</span> <span class="n">registry</span><span class="o">.</span><span class="n">extractor_names</span>
<span class="c1"># Returns: [&quot;BasicFileInfoExtractor&quot;, &quot;DM3Extractor&quot;, ..., &quot;TescanTiffExtractor&quot;]</span>
</pre></div>
</div>
</section>
<section id="extractor-modules">
<h3>Extractor Modules<a class="headerlink" href="#extractor-modules" title="Link to this heading">#</a></h3>
<p>For complete API documentation of the extractor modules, see:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.html#module-nexusLIMS.extractors" title="nexusLIMS.extractors"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors</span></code></a> - Main extractor module</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.registry.html#module-nexusLIMS.extractors.registry" title="nexusLIMS.extractors.registry"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.registry</span></code></a> - Extractor registry and auto-discovery</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.digital_micrograph.html#module-nexusLIMS.extractors.plugins.digital_micrograph" title="nexusLIMS.extractors.plugins.digital_micrograph"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.digital_micrograph</span></code></a> - DM3/DM4 file extractor</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.quanta_tif.html#module-nexusLIMS.extractors.plugins.quanta_tif" title="nexusLIMS.extractors.plugins.quanta_tif"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.quanta_tif</span></code></a> - FEI/Thermo TIF file extractor</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.orion_HIM_tif.html#module-nexusLIMS.extractors.plugins.orion_HIM_tif" title="nexusLIMS.extractors.plugins.orion_HIM_tif"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.orion_HIM_tif</span></code></a> - Zeiss Orion / Fibics HIM TIF file extractor</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.tescan_tif.html#module-nexusLIMS.extractors.plugins.tescan_tif" title="nexusLIMS.extractors.plugins.tescan_tif"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.tescan_tif</span></code></a> - Tescan PFIB/SEM TIF file extractor</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.fei_emi.html#module-nexusLIMS.extractors.plugins.fei_emi" title="nexusLIMS.extractors.plugins.fei_emi"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.fei_emi</span></code></a> - FEI TIA .ser/.emi file extractor</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.edax.html#module-nexusLIMS.extractors.plugins.edax" title="nexusLIMS.extractors.plugins.edax"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.edax</span></code></a> - EDAX .spc/.msa file extractor</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.basic_metadata.html#module-nexusLIMS.extractors.plugins.basic_metadata" title="nexusLIMS.extractors.plugins.basic_metadata"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.basic_metadata</span></code></a> - Basic metadata fallback extractor</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.plugins.preview_generators.html#module-nexusLIMS.extractors.plugins.preview_generators" title="nexusLIMS.extractors.plugins.preview_generators"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors.plugins.preview_generators</span></code></a> - Preview image generation utilities</p></li>
</ul>
</section>
</section>
<section id="further-reading">
<h2>Further Reading<a class="headerlink" href="#further-reading" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="record_building.html#record-building"><span class="std std-ref">Record building workflow</span></a> - How extractors fit into the record building workflow</p></li>
<li><p><a class="reference internal" href="taxonomy.html"><span class="doc">NexusLIMS Taxonomy</span></a> - Data type classification and taxonomy</p></li>
<li><p><a class="reference internal" href="../api/nexusLIMS/nexusLIMS.extractors.html#module-nexusLIMS.extractors" title="nexusLIMS.extractors"><code class="xref py py-mod docutils literal notranslate"><span class="pre">nexusLIMS.extractors</span></code></a> - Complete API documentation</p></li>
</ul>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="record_building.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Record building workflow</p>
      </div>
    </a>
    <a class="right-next"
       href="taxonomy.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">NexusLIMS Taxonomy</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quick-reference">Quick Reference</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#plugin-architecture">Plugin Architecture</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#fully-supported-formats">Fully Supported Formats</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#digital-micrograph-files-dm3-dm4">Digital Micrograph Files (.dm3, .dm4)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fei-thermo-fisher-tif-files-tif">FEI/Thermo Fisher TIF Files (.tif)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#zeiss-orion-fibics-him-tif-files-tif">Zeiss Orion / Fibics HIM TIF Files (.tif)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tescan-pfib-sem-tif-files-tif">Tescan PFIB/SEM TIF Files (.tif)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fei-tia-files-ser-emi">FEI TIA Files (.ser, .emi)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#edax-eds-files-spc-msa">EDAX EDS Files (.spc, .msa)</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#spc-files">.spc Files</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#msa-files">.msa Files</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#partially-supported-formats">Partially Supported Formats</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#image-formats">Image Formats</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#text-files-txt">Text Files (.txt)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#unsupported-formats">Unsupported Formats</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-extraction-works">How Extraction Works</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#file-discovery-and-strategy">File Discovery and Strategy</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extraction-process">Extraction Process</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#expected-metadata-structure">Expected Metadata Structure</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extension-fields-in-practice">Extension Fields in Practice</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#adding-support-for-new-formats">Adding Support for New Formats</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#api-reference">API Reference</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extractor-registry-properties">Extractor Registry Properties</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extractor-modules">Extractor Modules</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-reading">Further Reading</a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/user_guide/extractors.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, datasophos, LLC.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>